---
title: "Structural variations analysis with GALLO"
author: "Jenny Jakobsson"
date: "2/4/2021"
output: html_document

params: 
  workingDir: "/Users/jj/breedmaps/SVs_identification/"
  annDir: "data/annotation/"
  vcfDir: "data/vcf/"
  resultsDir: "results/GALLO/BTA125_S1_L001_R_SV/"
  gff_file: "results/reformatted_data/org_format_QTLdb.tsv"
  vcf_file: "BTA125_S1_L001_R_SV.vcf"
  gene_file: "Bos_taurus.ARS-UCD1.2.102.gtf"
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Install packages (only needs to be run 1 time)
#install.packages("devtools")
#install.packages("vcfR")
#install_github("pablobio/GALLO")
#install.packages("DT")

# Load the packages
library(devtools)
library(GALLO)
library(vcfR)
library("tidyverse")

```
# GALLO

GALLO is based on 8 functions and the QTLdb

1. import_gff_gtf(): Takes a .gft or .gff file and import into a dataframe
2. find_genes_qtls_around_markers: Takes a dataframe with candidate markers and/or regions (haplotypes, windows, CNVs, etc) and search for genes or QTLs in a specified interval
3. overlapping_among_groups: Takes a dataframe with a column for genes, QTLs (or any other data) and a grouping column and create matrices with the ovelapping information
4. plot_overlapping: Takes the output from overlapping_amoung_groups function and creates a heatmap with the overlapping between groups
5. plot_qtl_info: Takes the output from find_genes_qtls_around_markers and create plots for the frequency of each QTL type and trait
6. qtl_enrich: Takes the output from find_genes_qtls_around_markers and perform a QTL enrichment analysis
7. QTLenrich_plot: Takes the output from _find_genes_qtls_around_markers function and creates a heatmap with the overlapping between groups
8. relationship_plot: Takes the output from find_genes_qtls_around_markers function and creates a chord plot with the relationship between groups

# Load data
Gene information downloaded from ENSEMBL(ftp://ftp.ensembl.org/pub/release-102/gtf/bos_taurus/). QTL data downloaded from QTLdb(https://www.animalgenome.org/cgi-bin/QTLdb/BT/index) in gff format for reference UMD3.1.

```{r read files}
# Load VCF file
vcf_path = paste(params$workingDir, params$vcfDir, params$vcf_file, sep="")
vcf <- read.vcfR(vcf_path, verbose = FALSE)

# Load QTL info from QTLdb gff file
gff_path = paste(params$workingDir, params$gff_file, sep="")
annotation = import_gff_gtf(gff_path, "gff")

# Load gene info from ENSEMBL gtf file
gene_path = paste(params$workingDir, params$annDir, params$gene_file, sep="")
gene_ann = import_gff_gtf(gene_path, "gtf")

```



# Format data and filter

The files needs to be formatted to get the variant data. I filter out the variants that did not pass the quality control in DELLY.

```{r formatting and filtering}
# Convert vcfR object to a tidy object to more easily handle the data
df = vcfR2tidy(vcf, verbose=FALSE)

# Split the data frame. The variants are contained in the fix column
meta = data.frame(df$meta)
variants = data.frame(df$fix)
gt = data.frame(df$gt)

# Filter out variations that did not pass quality control
filt_variants = variants %>% filter(FILTER=="PASS")
# GALLO needs columns to be called CHR, BP1 and BP2.
filt_variants = filt_variants %>% dplyr::rename(CHR= "CHROM", BP1 = "POS", BP2="END")
```

# Find specific SVs
To more easily get an overview of the data it is split into the SV types. Breakends are not relevant here so only duplications, deletions and inversions are analyzed.

```{r remove break ends}
svs = filt_variants %>% filter(ALT == "<DUP>" | ALT == "<DEL>" | ALT == "<INV>")

```

# GALLO
Using GALLO the QTLs and genes around the variants can be found

```{r gallo}
svs_qtl = find_genes_qtls_around_markers(annotation,
                                         svs,
                                         method = "qtl",
                                         marker = "haplotype",
                                         interval = 0,
                                         nThreads = 1)
# Show subset of data
DT::datatable(svs_qtl[1:100,], rownames = FALSE, extensions = 'FixedColumns', options=list(scrollX = TRUE))

svs_gene = find_genes_qtls_around_markers(gene_ann,
                                          svs,
                                          method = "gene",
                                          marker = "haplotype")

# Show subset of data
DT::datatable(svs_gene[1:100,], rownames = FALSE, extensions = 'FixedColumns', options=list(scrollX = TRUE))

```
## Plotting

```{r plotting}
# Plots the distribution of the trait types
plot_qtl_info(svs_qtl, qtl_plot="qtl_type") %>% title(main="Structural variations", xlab="Trait percentage")
plot_qtl_info(svs_qtl, qtl_plot="qtl_name", n="all", qtl_class = "Milk")

# Performs a QTL enrichment analysis
enrich_qtl = qtl_enrich(annotation, svs_qtl, qtl_type="QTL_type", enrich_type="chromosome")
# Plot the enrichment
enrich_plot_qtl = QTLenrich_plot(enrich_qtl, "QTL", "pvalue")
enrich_plot_qtl
```

# Clean files
To make it easier to read the output some columns are removed.
```{r clean files}
# Clean QTL files
svs_qtl_cleaned = svs_qtl %>% dplyr::rename(
  SV_ID = "ID",
  SV_start = "BP1",
  SV_end = "BP2",
  QTL_chr = "chr",
  QTL_start = "start_pos",
  QTL_end = "end_pos",
  extra = "trait_ID",
  trait_ID = "breed"
) %>% dplyr::select(
  SV_ID,
  CHR,
  SV_start,
  SV_end,
  QTL_chr,
  QTL_start,
  QTL_end,
  QTL_type,
  QTL_ID,
  trait_ID,
  Name,
  pubmed_id,
  p_value
) %>% unique()


# Clean gene files
svs_gene_cleaned = svs_gene %>% dplyr::rename(
  SV_ID = "ID",
  SV_start = "BP1",
  SV_end = "BP2",
  gene_chr = "chr",
  gene_start = "start_pos",
  gene_end = "end_pos"
) %>% dplyr::select(
  SV_ID,
  CHR,
  SV_start,
  SV_end,
  gene_id,
  gene_name,
  gene_chr,
  gene_start,
  gene_end,
  gene_biotype
) %>% unique()

```



# Write to file

```{r print to file}
# Write QTL files
results_qtl = paste(params$workingDir,
                params$resultsDir,
                "svs_qtl.tsv",
                sep = "")
write.table(
  x = svs_qtl_cleaned,
  file = results_qtl,
  quote = F,
  sep = "\t",
  row.names = F,
  col.names = T
)

results_gene = paste(params$workingDir,
                params$resultsDir,
                "svs_gene.tsv",
                sep = "")
write.table(
  x = svs_gene_cleaned,
  file = results_gene,
  quote = F,
  sep = "\t",
  row.names = F,
  col.names = T
)


``` 




