---
title: "Gene annotation"
author: "Jenny Jakobsson"
date: "3/31/2021"
output: html_document

params:
  workingDir: '/Users/jj/breedmaps/SVs_identification/'
  annDir: 'data/annotation/'
  gene_ann: 'Bos_taurus.ARS-UCD1.2.102.gtf'
  vcfDir: 'data/vcf/'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("vcfR")

library(vcfR)
library("tidyverse")

```

## Load files
Gene information downloaded from ENSEMBL(ftp://ftp.ensembl.org/pub/release-102/gtf/bos_taurus/).

```{r read files}
# Load VCF file
vcf_path = paste(params$workingDir, params$vcfDir, sep = "")
vcf_files = list.files(path = vcf_path, pattern = "*.vcf")

vcf = read.table(
  file = paste(vcf_path, vcf_files[1], sep = ""),
  quote = "",
  sep = "\t",
  header = FALSE,
  stringsAsFactors = F
)
# fix = V8, Meta = V9, gt = V10
renamed = vcf %>% rename(
  CHROM = V1,
  POS = V2,
  ID = V3,
  REF = V4,
  ALT = V5,
  QUAL = V6,
  FILTER = V7,
  INFO = V8,
  FORMAT = V9,
  GT = V10
)
tmp.split = data.frame(stringr::str_split_fixed(renamed$INFO, ";", n = Inf))
info = data.frame(matrix(
  data = NA,
  nrow = dim(tmp.split)[1],
  ncol = dim(tmp.split)[2] + 1
))
rownames(info) = renamed$ID
colnames(info) = c(
  "PRECISE",
  "IMPRECISE",
  "SVTYPE",
  "SVMETHOD",
  "CHR2",
  "END",
  "PE",
  "MAPQ",
  "CT",
  "CIPOS",
  "CIEND",
  "SRMAPQ",
  "INSLEN",
  "HOMLEN",
  "SR",
  "SRQ",
  "CONSENSUS",
  "CE"
)
for (row in 1:dim(info)[1]) {
  for (col in 1:dim(info)[2]) {
    if (is.null(tmp.split[row, col])) {
      
    }
    else if (tmp.split[row, col] == "") {
      
    }
    else if (tmp.split[row, col] == "PRECISE") {
      info[row, "PRECISE"] = TRUE
      info[row, "IMPRECISE"] = FALSE
    }
    else if (tmp.split[row, col] == "IMPRECISE") {
      info[row, "PRECISE"] = FALSE
      info[row, "IMPRECISE"] = TRUE
    }
    else {
      split = stringr::str_split_fixed(tmp.split[row, col], "=", n = 2)
      #info[row, split[1]] = tmp.split[row, col]
      info[row, split[1]] = split[[2]]
    }
  }
}
id_column = rownames_to_column(info, var = "ID")
result = inner_join(renamed, id_column, by = "ID")
cleaned = result %>% select(CHROM, POS, ID, REF, ALT, QUAL, FILTER, PRECISE, SVTYPE, CHR2, END, PE, MAPQ, CT, CIPOS, CIEND, SRMAPQ, INSLEN, HOMLEN, SR, SRQ, CONSENSUS, CE)
qual = cleaned %>% filter(PRECISE==TRUE) %>% filter(FILTER=='PASS')

# for (i in 1:length(vcf_files)) {
#   vcf_file = paste(vcf_path, vcf_files[i], sep = "")
#   df = vcfR2tidy(read.vcfR(vcf_file, verbose = FALSE), verbose = FALSE)$fix
#   filt_df = df %>% filter(FILTER == "PASS")
#   vcf_range = makeGRangesFromDataFrame(
#     filt_df,
#     seqnames.field = "CHROM",
#     start.field = "POS",
#     end.field = "END",
#     strand.field = "K27ac_Strand"
#   )
#   assign(vcf_files[i], filt_df)
# }



# # Load gene info from ENSEMBL gtf file
# gene_path = paste(params$workingDir, params$annDir, params$gene_file, sep="")
# gene_ann = import_gff_gtf(gene_path, "gtf")
```



```{r formatting and filtering}



```



