---
title: "Structural variations analysis"
author: "Jenny Jakobsson"
date: "2/1/2021"
output: html_document

params: 
  workingDir: "/Users/jj/breedmaps/SVs_identification/"
  dataDir: "data/"
  gff_file: "UMD1.2_QTLdb_UTF-8.txt"
  vcf_file: "BTA125_S1_L001_R_SV.vcf"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("devtools")
#install.packages("vcfR")
#install_github("pablobio/GALLO")

library(devtools)
library(GALLO)
library(vcfR)
library("tidyverse")
library(GenomicRanges)

```

# Load data

```{r read files}
# Load VCF file
vcf_path = paste(params$workingDir, params$dataDir, params$vcf_file, sep="")
vcf <- read.vcfR(vcf_path)

# Load annotation gff file. Comes from QTLdb for UMD1.2 reference
gff_path = paste(params$workingDir, params$dataDir, params$gff_file, sep="")
annotation = import_gff_gtf(gff_path, "gff")
```

# Format data and filter

```{r formatting and filtering}
# Convert vcfR object to a tidy object
df = vcfR2tidy(vcf)

# Extract only the relevant data (not meta data and info columns)
meta = data.frame(df$meta)
variants = data.frame(df$fix)
gt = data.frame(df$gt)

#Filter out variations that did not pass quality control
filt_variants = variants %>% filter(FILTER=="PASS")
filt_variants = filt_variants %>% filter(ALT == "<DUP>" | ALT == "<DEL>" | ALT == "<INV>")
```


```{r fixing annotation}
# Creating field for the annotation
annotation_QTLID = annotation  %>% separate(
  col = extra_info,
  into = c("QTL_ID"),
  sep = ";",
  remove = TRUE
)
# annotation_clean = annotation  %>% separate(
#   col = extra_info,
#   into = c("QTL_ID", "Name","Abbrev","PUBMED_ID", "trait_ID","trait", "trait_type","FlankMarker","breed","extra1", "extra2", "extra3"),
#   sep = ";",
#   remove = TRUE
# )
# annotation2 = annotation_clean  %>% separate(
#   col = breed,
#   into = c("other", "breed"),
#   sep=c("breed="),
#   remove=TRUE
# )
# annotation3 = annotation2 %>% separate(
#   col = 
# )
# annotation_final = annotation3 %>% dplyr::select("QTL_ID", "Name","Abbrev","PUBMED_ID", "trait_ID","trait", "trait_type","FlankMarker","breed")

```

## Add annotation for the genes

```{r adding annotation}
annotation_range = makeGRangesFromDataFrame(
  annotation_QTLID,
  seqnames.field = "chr",
  start.field = "start_pos",
  end.field = "end_pos"
)
variant_range = makeGRangesFromDataFrame(
  filt_variants,
  seqnames.field = "CHROM",
  start.field = "POS",
  end.field = "END"
  
)
overlap = data.frame(findOverlaps(variant_range, annotation_range, maxgap = 0))
overlap$ID = filt_variants$ID[overlap$queryHits]
overlap$QTL_ID = annotation_QTLID$QTL_ID[overlap$subjectHits]
overlap = overlap %>% dplyr::select(ID, QTL_ID)
full_df = inner_join(filt_variants, overlap, by="ID")
full_df2 = inner_join(full_df, annotation_QTLID, by="QTL_ID")
full_unique = unique(full_df2)

joined_data = full_unique %>% dplyr::select(CHROM, ID, QTL_ID, QTL_type, POS, END, start_pos, end_pos)
```

# Find specific SVs

```{r find specific SVs}
# duplications = filt_variants %>% filter(ALT=="<DUP>")
# deletions = filt_variants %>% filter(ALT=="<DEL>")
# inversions = filt_variants %>% filter(ALT=="<INV>")
# breakends = filt_variants %>% filter(ALT != "<DUP>" & ALT != "<DEL>" & ALT != "<INV>")

```

# Results
## BTA125_S1
L001: 297 deletions,
	40 duplications,
	56 inversions,
	113 breakends.

L002: 307 deletions, 
	42 duplications, 
	45 inversions, 
	114 breakends.

L003: 269 deletions, 
	34 duplications, 
	48 inversions, 
	108 breakends.




