---
title: "Gene analysis"
author: "Jenny Jakobsson"
date: "2/16/2021"
output: html_document

params: 
  workingDir: "/Users/jj/breedmaps/SVs_identification/"
  annDir: "data/annotation/"
  vcfDir: "data/vcf/"
  vcf_file: "BTA125_S1_L001_R_SV.vcf"
  gene_file: "Bos_taurus.ARS-UCD1.2.102.gtf"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Install packages (only needs to be run 1 time)
#install.packages("BiocManager")
#install.packages("tidyverse")
#BiocManager::install("rtracklayer")

library("tidyverse")
library("rtracklayer")

```

```{r load the files}
# Load VCF file
vcf_path = paste(params$workingDir, params$vcfDir, params$vcf_file, sep="")
vcf = read.table(
  file = vcf_path ,
  quote = "",
  sep = "\t",
  header = FALSE,
  stringsAsFactors = FALSE
)
# fix = V8, Meta = V9, gt = V10
renamed = vcf %>% rename(CHROM=V1, POS=V2, ID=V3, REF=V4, ALT=V5, QUAL = V6, FILTER=V7, INFO=V8, FORMAT=V9, GT=V10)
tmp.split = data.frame(stringr::str_split_fixed(renamed$INFO, ";",n=Inf))
info = data.frame(matrix(data=NA, nrow=dim(tmp.split)[1], ncol=dim(tmp.split)[2]+1 ))
rownames(info) = renamed$ID
colnames(info) = c("PRECISE", "IMPRECISE", "SVTYPE", "SVMETHOD", "CHR2", "END", "PE", "MAPQ", "CT", "CIPOS", "CIEND","SRMAPQ", "INSLEN", "HOMLEN", "SR", "SRQ", "CONSENSUS", "CE")
for (row in 1:dim(info)[1]) {
  for (col in 1:dim(info)[2]) {
    if (is.null(tmp.split[row, col])) {
      
    }
    else if (tmp.split[row, col] == "") {
      
    }
    else if (tmp.split[row, col] == "PRECISE") {
      info[row, "PRECISE"] = TRUE
      info[row, "IMPRECISE"] = FALSE
    }
    else if (tmp.split[row, col] == "IMPRECISE") {
      info[row, "PRECISE"] = FALSE
      info[row, "IMPRECISE"] = TRUE
    }
    else {
      split = stringr::str_split_fixed(tmp.split[row, col], "=", n = 2)
      info[row, split[1]] = tmp.split[row, col]
    }
  }
}
id_column = rownames_to_column(info, var="ID")
result = inner_join(renamed,id_column, by="ID")
#precise = info %>% filter(PRECISE==TRUE)
```

```{r extract variants}
# Load gene info from ENSEMBL gtf file
gene_path = paste(params$workingDir, params$annDir, params$gene_file, sep="")
annotation = read.table(
  file = gene_path ,
  quote = "",
  sep = "\t",
  header = FALSE,
  stringsAsFactors = FALSE
)
gtf.file<-rtracklayer::import(gene_path)
a = as.data.frame(gtf.file)
a = a %>% filter(type == "gene")

```


