---
title: "Gene annotation"
author: "Jenny Jakobsson"
date: "3/31/2021"
output: html_document

params:
  workingDir: '/Users/jj/breedmaps/SVs_identification/'
  annDir: 'data/annotation/'
  scriptDir: 'scripts/gene_annotation/'
  gene_ann: 'Bos_taurus.ARS-UCD1.2.103.gtf'
  vcfDir: 'data/vcf/'
  resultsDir: 'results/gene_annotation/'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("UniprotR")

library("tidyverse")
library(GenomicRanges)
library("rtracklayer")
source(paste(params$workingDir, params$scriptDir, "functions.R", sep=""))
library('UniprotR')
```


## Load files
Gene information downloaded from ENSEMBL(ftp://ftp.ensembl.org/pub/release-102/gtf/bos_taurus/).

```{r load genes}
# Load gene info from ENSEMBL gtf file
gene_path = paste(params$workingDir, params$annDir, params$gene_ann, sep =
                    "")
gtf.file <- rtracklayer::import(gene_path)
gene_df = as.data.frame(gtf.file) %>% filter(type == "gene") %>% select(seqnames, start, end, width, strand, source, type, gene_id, gene_source, gene_biotype, gene_name)
gene_range = makeGRangesFromDataFrame(
  gene_df,
  seqnames.field = "seqnames",
  start.field = "start",
  end.field = "end",
  strand.field = "strand"
)

```

```{r read files}
# Load VCF file
vcf_path = paste(params$workingDir, params$vcfDir, sep = "")
vcf_files = list.files(path = vcf_path, pattern = "*.vcf")
vcf_names = list()
df_list = list()
range_list = list()

for (i in 1:length(vcf_files)) {
  vcf_file = paste(vcf_path, vcf_files[i], sep = "")
  
  # Save the file name to be able to use it when writing results files
  name = strsplit(vcf_files, "\\.")[[i]][1]
  vcf_names[[i]] = name
  
  df = read_delly_vcf(vcf_file)
  # Because of the ranges need to filter out the breakpoint
  filt_var = df %>% filter(ALT == "<DUP>" | ALT == "<DEL>" |
                             ALT == "<INV>" | ALT == "<INS>")
  df_list[[i]] = filt_var
  vcf_range = makeGRangesFromDataFrame(
    filt_var,
    seqnames.field = "CHROM",
    start.field = "POS",
    end.field = "END",
  )
  range_list[[i]] = vcf_range
}

```


```{r finding overlap}
overlap_list = list()
for (i in 1:length(range_list)) {
  overlap = findoverlap_dataframe(range_list[[i]], gene_range, df_list[[i]], gene_df)
  precise_only = overlap %>% filter(PRECISE == TRUE) %>% filter(FILTER == "PASS")
  overlap_list[[i]] = overlap
}
```


```{r filtering}
filtered_list = list()
for (i in 1:length(range_list)) {
  filtered_list[[i]] = overlap_list[[i]] %>% filter(FILTER == "PASS") %>% filter(PRECISE == TRUE)
}
```


```{r write to file}
# Save all variants with genes to file
for (i in 1:length(overlap_list)) {
  results_gene = paste(params$workingDir,
                       params$resultsDir,
                       "all_variants/",
                       "ann_",
                       vcf_names[[i]],
                       ".tsv",
                       sep = "")
  write.table(
    x = overlap_list[[i]],
    file = results_gene,
    quote = F,
    sep = "\t",
    row.names = F,
    col.names = T
  )
}

# Save only the PRECISE variants
for (i in 1:length(overlap_list)) {
  results_gene = paste(
    params$workingDir,
    params$resultsDir,
    "filtered_variants/",
    "filt_ann_",
    vcf_names[[i]],
    ".tsv",
    sep = ""
  )
  write.table(
    x = filtered_list[[i]],
    file = results_gene,
    quote = F,
    sep = "\t",
    row.names = F,
    col.names = T
  )
}

for (i in 1:length(overlap_list)) {
  results_uniprot = paste(
    params$workingDir,
    params$resultsDir,
    "filtered_variants/",
    "UNIPROT_",
    vcf_names[[i]],
    ".csv",
    sep = ""
  )
  uniprot_genes = data.frame(acc = filtered_list[[i]]$gene_id)
  write.csv(uniprot_genes,results_uniprot, row.names = FALSE)
}

```



