---
title: "Gene annotation"
author: "Jenny Jakobsson"
date: "3/31/2021"
output: html_document

params:
  workingDir: '/Users/jj/breedmaps/SVs_identification/'
  annDir: 'data/annotation/'
  scriptDir: 'scripts/gene_annotation/'
  gene_ann: 'Bos_taurus.ARS-UCD1.2.103.gtf'
  vcfDir: 'data/vcf/'
  resultsDir: 'results/gene_annotation/'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("UniprotR")

library("tidyverse")
library(GenomicRanges)
library("rtracklayer")
source(paste(params$workingDir, params$scriptDir, "functions.R", sep=""))
library('UniprotR')
```


## Load files
Gene information downloaded from ENSEMBL(ftp://ftp.ensembl.org/pub/release-102/gtf/bos_taurus/).
This is a power heavy script so it needs to be run in parts. It is divided into: annotation of genes, exon, transcripts and the rest. The result is therfore divided into four different result files.

```{r read vcf files}
# Load VCF file
vcf_path = paste(params$workingDir, params$vcfDir, sep = "")
vcf_files = list.files(path = vcf_path, pattern = "*.vcf")
vcf_names = list()
df_list = list()
range_list = list()

for (i in 1:length(vcf_files)) {
  vcf_file = paste(vcf_path, vcf_files[i], sep = "")
  
  # Save the file name to be able to use it when writing results files
  name = strsplit(vcf_files, "\\.")[[i]][1]
  vcf_names[[i]] = name
  
  df = read_delly_vcf(vcf_file)
  # Because of the ranges need to filter out the breakpoint
  filt_var = df %>% dplyr::filter(ALT == "<DUP>" | ALT == "<DEL>" |
                                    ALT == "<INV>" | ALT == "<INS>")
  filt_var = filt_var  %>% dplyr::filter(FILTER == "PASS") %>% filter(PRECISE == TRUE)
  filt_var_final = filt_var %>% dplyr::mutate(SV_LENGTH = as.numeric(END) -
                                                as.numeric(POS)) %>% dplyr::rename(ID1 = ID)
  df_list[[i]] = filt_var_final
  vcf_range = makeGRangesFromDataFrame(
    filt_var_final,
    seqnames.field = "CHROM",
    start.field = "POS",
    end.field = "END",
  )
  range_list[[i]] = vcf_range
}

```

```{r load annotation}
# Load gene info from ENSEMBL gtf file
ann_path = paste(params$workingDir, params$annDir, params$gene_ann, sep =
                    "")
ann_df = as.data.frame(rtracklayer::import(ann_path))
ann_df = ann_df %>% dplyr::rename(gene_chrom = seqnames, gene_start = start, gene_end = end, gene_strand = strand, gene_width = width, gene_type = type) %>% dplyr::select(-source)
```

```{r get gene ann}
# The overlap function requires that the ID for the two overlapping data sets have ID1 or ID2 for the ID
gene_df = ann_df %>% filter(gene_type == "gene") %>% dplyr::rename(ID2 = gene_id)
gene_range = makeGRangesFromDataFrame(
  gene_df,
  seqnames.field = "gene_chrom",
  start.field = "gene_start",
  end.field = "gene_end",
  strand.field = "gene_strand"
)

overlap_gene_list = list()
for (i in 1:length(range_list)) {
  overlap_gene = findoverlap_dataframe(range_list[[i]], gene_range, df_list[[i]], gene_df)
  overlap_gene_list[[i]] = overlap_gene %>% dplyr::rename(ID = ID1, gene_id = ID2)
}
```

```{r get transcript ann}
# The overlap function requires that the ID for the two overlapping data sets have ID1 or ID2 for the ID
tran_df = ann_df %>% filter(gene_type == "transcript") %>% dplyr::rename(ID2 = gene_id)
tran_range = makeGRangesFromDataFrame(
  tran_df,
  seqnames.field = "gene_chrom",
  start.field = "gene_start",
  end.field = "gene_end",
  strand.field = "gene_strand"
)

overlap_tran_list = list()
for (i in 1:length(range_list)) {
  overlap_tran = findoverlap_dataframe(range_list[[i]], tran_range, df_list[[i]], tran_df)
  overlap_tran_list[[i]] = overlap_tran %>% dplyr::rename(ID = ID1, gene_id = ID2)
}
```

```{r get exon ann}
exon_df = ann_df %>% filter(gene_type == "exon") %>% dplyr::rename(ID2 = gene_id)
exon_range = makeGRangesFromDataFrame(
  exon_df,
  seqnames.field = "gene_chrom",
  start.field = "gene_start",
  end.field = "gene_end",
  strand.field = "gene_strand"
)

overlap_exon_list = list()
for (i in 1:length(range_list)) {
  overlap_exon = findoverlap_dataframe(range_list[[i]], exon_range, df_list[[i]], exon_df)
  overlap_exon_list[[i]] = overlap_exon %>% dplyr::rename(ID = ID1, gene_id = ID2)
}
```

```{r get the rest of the ann}
rest_df = ann_df %>% filter(gene_type != "exon") %>% filter(gene_type != "gene") %>% filter(gene_type != "transcript") %>% dplyr::rename(ID2 = gene_id)
rest_range = makeGRangesFromDataFrame(
  rest_df,
  seqnames.field = "gene_chrom",
  start.field = "gene_start",
  end.field = "gene_end",
  strand.field = "gene_strand"
)

overlap_rest_list = list()
for (i in 1:length(range_list)) {
  overlap_rest = findoverlap_dataframe(range_list[[i]], rest_range, df_list[[i]], rest_df)
  overlap_rest_list[[i]] = overlap_rest %>% dplyr::rename(ID = ID1, gene_id = ID2)
}
```


```{r write to file}

# Save only the PRECISE variants
for (i in 1:length(range_list)) {
  filt_results = paste(
    params$workingDir,
    params$resultsDir,
    "filtered_variants/",
    "filt_ann_gene_",
    vcf_names[[i]],
    ".tsv",
    sep = ""
  )
  write.table(
    x = overlap_gene_list[[i]],
    file = filt_results,
    quote = F,
    sep = "\t",
    row.names = F,
    col.names = T
  )
  
  filt_results = paste(
    params$workingDir,
    params$resultsDir,
    "filtered_variants/",
    "filt_ann_exon_",
    vcf_names[[i]],
    ".tsv",
    sep = ""
  )
  write.table(
    x = overlap_exon_list[[i]],
    file = filt_results,
    quote = F,
    sep = "\t",
    row.names = F,
    col.names = T
  )
  
  filt_results = paste(
    params$workingDir,
    params$resultsDir,
    "filtered_variants/",
    "filt_ann_transcript_",
    vcf_names[[i]],
    ".tsv",
    sep = ""
  )
  write.table(
    x = overlap_tran_list[[i]],
    file = filt_results,
    quote = F,
    sep = "\t",
    row.names = F,
    col.names = T
  )
  filt_results = paste(
    params$workingDir,
    params$resultsDir,
    "filtered_variants/",
    "filt_ann_rest_",
    vcf_names[[i]],
    ".tsv",
    sep = ""
  )
  write.table(
    x = overlap_rest_list[[i]],
    file = filt_results,
    quote = F,
    sep = "\t",
    row.names = F,
    col.names = T
  )
}

```



